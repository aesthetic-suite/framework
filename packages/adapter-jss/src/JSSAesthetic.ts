import Aesthetic, {
  AestheticOptions,
  ClassName,
  Ruleset,
  Sheet,
  StyleName,
  SheetMap,
} from 'aesthetic';
import { JSS, StyleSheet as JSSSheet } from 'jss';
import { NativeBlock, ParsedBlock } from './types';

export default class JSSAesthetic<Theme extends object> extends Aesthetic<
  Theme,
  NativeBlock,
  ParsedBlock
> {
  jss: JSS;

  keyframes: { [animationName: string]: string } = {};

  sheets: { [styleName: string]: JSSSheet<any> } = {};

  constructor(jss: JSS, options: Partial<AestheticOptions> = {}) {
    super(options);

    this.jss = jss;

    this.syntax
      .on('attribute', this.handleNested)
      .on('charset', this.handleCharset)
      .on('css', this.handleCss)
      .on('fallback', this.handleFallback)
      .on('font-face', this.handleFontFace)
      .on('global', this.handleGlobal)
      .on('import', this.handleImport)
      .on('keyframe', this.handleKeyframe)
      .on('media', this.handleMedia)
      .on('property', this.handleProperty)
      .on('pseudo', this.handleNested)
      .on('selector', this.handleNested)
      .on('support', this.handleSupport)
      .on('viewport', this.handleViewport);
  }

  flushStyles(styleName: StyleName) {
    const sheet = this.sheets[styleName];

    if (sheet) {
      sheet.attach();
    }
  }

  isParsedBlock(block: ParsedBlock): boolean {
    return typeof block === 'string';
  }

  protected processStyleSheet(
    styleSheet: SheetMap<NativeBlock>,
    styleName: StyleName,
  ): SheetMap<ParsedBlock> {
    this.sheets[styleName] = this.jss.createStyleSheet<any>(styleSheet, {
      media: 'screen',
      meta: styleName,
    });

    return this.sheets[styleName].classes;
  }

  protected transformToClassName(styles: ParsedBlock[]): ClassName {
    return styles.join(' ');
  }

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleCharset = (sheet: Sheet<NativeBlock>, charset: string) => {
    sheet.addAtRule('@charset', charset);
  };

  private handleCss = (css: string) => {
    this.getStyleSheetManager().injectStatements(css);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#fallbacks
  private handleFallback = (
    ruleset: Ruleset<NativeBlock>,
    name: keyof NativeBlock,
    value: any[],
  ) => {
    const fallbacks = value.map(fallback => ({ [name]: fallback }));
    const current = ruleset.properties.fallbacks || [];

    ruleset.addProperty('fallbacks', [...current, ...fallbacks]);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#font-face
  private handleFontFace = (sheet: Sheet<NativeBlock>, fontFaces: Ruleset<NativeBlock>[]) => {
    const current = (sheet.atRules['@font-face'] as Ruleset<NativeBlock>[]) || [];

    sheet.addAtRule('@font-face', [...current, ...fontFaces]);
  };

  // https://github.com/cssinjs/jss-global
  private handleGlobal = (
    sheet: Sheet<NativeBlock>,
    selector: string,
    ruleset: Ruleset<NativeBlock>,
  ) => {
    const current = (sheet.atRules['@global'] as Sheet<NativeBlock>) || sheet.createSheet();

    current.addRuleset(ruleset);

    sheet.addAtRule('@global', current);
  };

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleImport = (sheet: Sheet<NativeBlock>, paths: string[]) => {
    sheet.addAtRule('@import', paths);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#keyframes-animation
  private handleKeyframe = (
    sheet: Sheet<NativeBlock>,
    keyframe: Ruleset<NativeBlock>,
    animationName: string,
  ) => {
    let name = this.keyframes[animationName];

    // Used to avoid global collision
    if (!name) {
      name = `${animationName}-${Object.keys(this.keyframes).length}`;

      this.keyframes[animationName] = name;
    }

    sheet.addAtRule(`@keyframes ${name}`, keyframe);
  };

  // https://github.com/cssinjs/jss-nested#use-at-rules-inside-of-regular-rules
  private handleMedia = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`@media ${query}`, value);
  };

  // https://github.com/cssinjs/jss-nested#use--to-reference-selector-of-the-parent-rule
  private handleNested = (
    ruleset: Ruleset<NativeBlock>,
    selector: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`&${selector}`, value);
  };

  // https://github.com/cssinjs/jss-nested#use--to-reference-selector-of-the-parent-rule
  private handleProperty = (ruleset: Ruleset<NativeBlock>, name: keyof NativeBlock, value: any) => {
    if (name === 'animationName') {
      ruleset.addProperty(name, this.syntax.injectKeyframes(value, this.keyframes).join(', '));
    } else {
      ruleset.addProperty(name, value);
    }
  };

  // https://github.com/cssinjs/jss-nested#use-at-rules-inside-of-regular-rules
  private handleSupport = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`@supports ${query}`, value);
  };

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleViewport = (sheet: Sheet<NativeBlock>, ruleset: Ruleset<NativeBlock>) => {
    sheet.addAtRule('@viewport', ruleset);
  };
}
