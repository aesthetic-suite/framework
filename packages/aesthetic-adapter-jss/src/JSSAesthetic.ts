/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Aesthetic, {
  AestheticOptions,
  ClassName,
  Keyframes,
  Ruleset,
  Sheet,
  StyleName,
  StyleSheetMap,
} from 'aesthetic';
import { JSS, StyleSheet as JSSSheet } from 'jss';
import { NativeBlock, ParsedBlock } from './types';

export default class JSSAdapter<Theme> extends Aesthetic<Theme, NativeBlock, ParsedBlock> {
  jss: JSS;

  sheets: { [styleName: string]: JSSSheet<any> } = {};

  constructor(jss: JSS, options: Partial<AestheticOptions> = {}) {
    super(options);

    this.jss = jss;

    this.syntax
      .on('attribute', this.handleNested)
      .on('charset', this.handleCharset)
      .on('fallback', this.handleFallback)
      .on('font-face', this.handleFontFace)
      .on('global', this.handleGlobal)
      .on('import', this.handleImport)
      .on('keyframe', this.handleKeyframe)
      .on('media', this.handleMedia)
      .on('property', this.handleProperty)
      .on('pseudo', this.handleNested)
      .on('selector', this.handleNested)
      .on('support', this.handleSupport)
      .on('viewport', this.handleViewport);
  }

  protected processStyleSheet(
    styleSheet: object,
    styleName: StyleName,
  ): StyleSheetMap<ParsedBlock> {
    this.sheets[styleName] = this.jss
      .createStyleSheet<any>(styleSheet, {
        media: 'screen',
        meta: styleName,
        classNamePrefix: `${styleName}-`,
      })
      .attach();

    return this.sheets[styleName].classes;
  }

  protected transformToClassName(styles: (NativeBlock | ParsedBlock)[]): ClassName {
    const legitStyles: ParsedBlock[] = [];
    const tempStylesheet: { [key: string]: NativeBlock } = {};
    let counter = 0;

    styles.forEach(style => {
      if (typeof style === 'string') {
        legitStyles.push(style);
      } else if (typeof style === 'object' && style !== null) {
        tempStylesheet[`inline-${counter}`] = style;
        counter += 1;
      }
    });

    if (counter > 0) {
      legitStyles.push(...Object.values(this.processStyleSheet(tempStylesheet, 'inline-dynamic')));
    }

    return legitStyles.join(' ');
  }

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleCharset = (sheet: Sheet<NativeBlock>, charset: string) => {
    sheet.addAtRule('@charset', charset);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#fallbacks
  private handleFallback = (
    ruleset: Ruleset<NativeBlock>,
    name: keyof NativeBlock,
    value: any[],
  ) => {
    const fallbacks = value.map(fallback => ({ [name]: fallback }));
    const current = ruleset.properties.fallbacks || [];

    ruleset.addProperty('fallbacks', [...current, ...fallbacks]);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#font-face
  private handleFontFace = (sheet: Sheet<NativeBlock>, fontFaces: NativeBlock[]) => {
    const current = (sheet.atRules['@font-face'] as NativeBlock[]) || [];

    sheet.addAtRule('@font-face', [...current, ...fontFaces]);
  };

  // https://github.com/cssinjs/jss-global
  private handleGlobal = (
    sheet: Sheet<NativeBlock>,
    selector: string,
    ruleset: Ruleset<NativeBlock>,
  ) => {
    const current = (sheet.atRules['@global'] as Sheet<NativeBlock>) || new Sheet<NativeBlock>();

    current.addRuleset(ruleset);

    sheet.addAtRule('@global', current);
  };

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleImport = (sheet: Sheet<NativeBlock>, paths: string[]) => {
    sheet.addAtRule('@import', paths);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#keyframes-animation
  private handleKeyframe = (
    sheet: Sheet<NativeBlock>,
    keyframes: Keyframes<NativeBlock>,
    animationName: string,
  ) => {
    sheet.addAtRule(`@keyframes ${animationName}`, keyframes);
  };

  // https://github.com/cssinjs/jss-nested#use-at-rules-inside-of-regular-rules
  private handleMedia = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`@media ${query}`, value);
  };

  // https://github.com/cssinjs/jss-nested#use--to-reference-selector-of-the-parent-rule
  private handleNested = (
    ruleset: Ruleset<NativeBlock>,
    selector: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`&${selector}`, value);
  };

  // https://github.com/cssinjs/jss-nested#use--to-reference-selector-of-the-parent-rule
  private handleProperty = (ruleset: Ruleset<NativeBlock>, name: keyof NativeBlock, value: any) => {
    if (name === 'animationName') {
      ruleset.addProperty(name, `$${value}`);
    } else {
      ruleset.addProperty(name, value);
    }
  };

  // https://github.com/cssinjs/jss-nested#use-at-rules-inside-of-regular-rules
  private handleSupport = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`@supports ${query}`, value);
  };

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleViewport = (sheet: Sheet<NativeBlock>, ruleset: Ruleset<NativeBlock>) => {
    sheet.addAtRule('@viewport', ruleset);
  };
}
