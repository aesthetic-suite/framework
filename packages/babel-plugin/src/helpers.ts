import { types as t } from '@babel/core';
import { createDebugger } from '@boost/debug';

export const debug = createDebugger('aesthetic:babel');

export function isFunction(node: t.Node) {
  return t.isFunctionExpression(node) || t.isArrowFunctionExpression(node);
}
