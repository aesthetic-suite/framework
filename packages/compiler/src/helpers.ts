import { ColorShade } from '@aesthetic/system';

export function formatShade(value: number | string): ColorShade {
	return String(value).padStart(2, '0').slice(0, 2) as ColorShade;
}

export function formatUnit(value: number): string {
	return value.toFixed(2).replace('.00', '');
}
