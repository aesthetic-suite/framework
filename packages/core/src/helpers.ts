import { Theme } from '@aesthetic/system';
import { objectLoop } from '@aesthetic/utils';
import { SheetParams } from './types';

export function createCacheKey(params: SheetParams, type: string): string {
  let key = type;

  // Since all other values are scalars, we can just join the values.
  // This is 3x faster than JSON.stringify(), and 1.5x faster than Object.values()!
  objectLoop(params, (value) => {
    key += value;
  });

  return key;
}

export function createDefaultParams(theme: Theme<{}>, params: SheetParams): Required<SheetParams> {
  return {
    contrast: theme.contrast,
    deterministic: false,
    direction: 'ltr',
    scheme: theme.scheme,
    theme: theme.name,
    unit: 'px',
    vendor: false,
    ...params,
  };
}
