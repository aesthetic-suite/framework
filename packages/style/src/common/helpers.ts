/* eslint-disable no-magic-numbers */

import { Value, MaybeValue, CSS, Sheet } from '@aesthetic/types';
import { hyphenate } from '@aesthetic/utils';
import { IMPORT_RULE, STYLE_RULE } from './constants';

export function insertRule(sheet: Sheet, rule: CSS, index?: number): number {
  try {
    return sheet.insertRule(rule, index ?? sheet.cssRules.length);
  } catch {
    // Vendor prefixed properties, pseudos, etc, that are inserted
    // into different vendors will trigger a failure. For example,
    // `-moz` or `-ms` being inserted into WebKit.
    // There's no easy way around this, so let's just ignore the
    // error so that subsequent styles are inserted.
    // istanbul ignore next
    return -1;
  }
}

export function insertAtRule(sheet: Sheet, rule: CSS): number {
  const { length } = sheet.cssRules;
  let index = 0;

  // At-rules must be inserted before normal style rules.
  for (let i = 0; i <= length; i += 1) {
    index = i;

    if (sheet.cssRules[i]?.type === STYLE_RULE) {
      break;
    }
  }

  return insertRule(sheet, rule, index);
}

export function insertImportRule(sheet: Sheet, rule: CSS): number {
  const { length } = sheet.cssRules;
  let index = 0;

  // Import rules must be inserted at the top of the style sheet,
  // but we also want to persist the existing order.
  for (let i = 0; i <= length; i += 1) {
    index = i;

    if (sheet.cssRules[i]?.type !== IMPORT_RULE) {
      break;
    }
  }

  return insertRule(sheet, rule, index);
}

export function isAtRule(value: string): boolean {
  return value[0] === '@';
}

export function isImportRule(value: string): boolean {
  return value.slice(0, 7) === '@import';
}

export function isMediaRule(value: string): boolean {
  return value.slice(0, 6) === '@media';
}

export function isNestedSelector(value: string): boolean {
  const char = value[0];

  return (
    char === ':' ||
    char === '[' ||
    char === '>' ||
    char === '~' ||
    char === '+' ||
    char === '*' ||
    char === '|'
  );
}

export function isSupportsRule(value: string): boolean {
  return value.slice(0, 9) === '@supports';
}

const unitlessProperties = new Set<string>();

[
  'animationIterationCount',
  'borderImage',
  'borderImageOutset',
  'borderImageSlice',
  'borderImageWidth',
  'columnCount',
  'columns',
  'flex',
  'flexGrow',
  'flexPositive',
  'flexShrink',
  'flexNegative',
  'flexOrder',
  'fontWeight',
  'gridArea',
  'gridRow',
  'gridRowEnd',
  'gridRowSpan',
  'gridRowStart',
  'gridColumn',
  'gridColumnEnd',
  'gridColumnSpan',
  'gridColumnStart',
  'lineClamp',
  'lineHeight',
  'maskBorder',
  'maskBorderOutset',
  'maskBorderSlice',
  'maskBorderWidth',
  'opacity',
  'order',
  'orphans',
  'tabSize',
  'widows',
  'zIndex',
  'zoom',
  // SVG
  'fillOpacity',
  'floodOpacity',
  'stopOpacity',
  'strokeDasharray',
  'strokeDashoffset',
  'strokeMiterlimit',
  'strokeOpacity',
  'strokeWidth',
].forEach((property) => {
  unitlessProperties.add(property);
  unitlessProperties.add(hyphenate(property));
});

export function isUnitlessProperty(property: string): boolean {
  return unitlessProperties.has(property);
}

export function isValidValue(property: string, value: MaybeValue | boolean): value is Value {
  if (value === null || value === undefined || value === true || value === false || value === '') {
    if (__DEV__) {
      // eslint-disable-next-line no-console
      console.warn(`Invalid value "${value}" for "${property}".`);
    }

    return false;
  }

  return true;
}

export function isVariable(value: string): boolean {
  return value.slice(0, 2) === '--';
}

export function joinQueries(prev: string | undefined, next: string): string {
  if (prev) {
    return `${prev} and ${next}`;
  }

  return next;
}
