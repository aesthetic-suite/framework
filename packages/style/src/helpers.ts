/* eslint-disable no-magic-numbers */

import { Value, MaybeValue } from '@aesthetic/types';
import { hyphenate } from '@aesthetic/utils';

export function isAtRule(value: string): boolean {
  return value[0] === '@';
}

export function isImportRule(value: string): boolean {
  return value.slice(0, 7) === '@import' && value[7] === ' ';
}

export function isNestedSelector(value: string): boolean {
  const char = value[0];

  return (
    char === ':' ||
    char === '[' ||
    char === '>' ||
    char === '~' ||
    char === '+' ||
    char === '*' ||
    char === '|'
  );
}

const unitlessProperties = new Set<string>();

[
  'animationIterationCount',
  'borderImage',
  'borderImageOutset',
  'borderImageSlice',
  'borderImageWidth',
  'columnCount',
  'columns',
  'flex',
  'flexGrow',
  'flexPositive',
  'flexShrink',
  'flexNegative',
  'flexOrder',
  'fontWeight',
  'gridArea',
  'gridRow',
  'gridRowEnd',
  'gridRowSpan',
  'gridRowStart',
  'gridColumn',
  'gridColumnEnd',
  'gridColumnSpan',
  'gridColumnStart',
  'lineClamp',
  'lineHeight',
  'maskBorder',
  'maskBorderOutset',
  'maskBorderSlice',
  'maskBorderWidth',
  'opacity',
  'order',
  'orphans',
  'tabSize',
  'widows',
  'zIndex',
  'zoom',
  // SVG
  'fillOpacity',
  'floodOpacity',
  'stopOpacity',
  'strokeDasharray',
  'strokeDashoffset',
  'strokeMiterlimit',
  'strokeOpacity',
  'strokeWidth',
].forEach((property) => {
  unitlessProperties.add(property);
  unitlessProperties.add(hyphenate(property));
});

export function isUnitlessProperty(property: string): boolean {
  return unitlessProperties.has(property);
}

export function isValidValue(property: string, value: MaybeValue | boolean): value is Value {
  if (value === null || value === undefined || value === true || value === false || value === '') {
    if (__DEV__) {
      console.warn(`Invalid value "${value}" for "${property}".`);
    }

    return false;
  }

  return true;
}

export function isVariable(value: string): boolean {
  return value.slice(0, 2) === '--';
}
