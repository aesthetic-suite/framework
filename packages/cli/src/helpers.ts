import { CONFIG_FOLDER } from '@aesthetic/compiler';
import { Path } from '@boost/common';
import { CWD } from './constants';

export function getConfigFolderDir(name: string): Path {
  return new Path(CWD, CONFIG_FOLDER, name);
}

export function validateSystemName(value: string) {
  if (value && !value.match(/^[-a-z0-9]+$/giu)) {
    throw new Error('Name may only contain numbers, letters, and dashes.');
  }
}
