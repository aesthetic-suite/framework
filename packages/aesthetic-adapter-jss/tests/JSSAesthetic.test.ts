import { create } from 'jss';
import preset from 'jss-preset-default';
import JSSAesthetic from '../src/JSSAesthetic';
import {
  FONT_ROBOTO_FLAT_SRC,
  FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
  KEYFRAME_FADE,
  SYNTAX_UNIFIED_LOCAL_FULL,
  SYNTAX_ATTRIBUTE,
  SYNTAX_DESCENDANT,
  SYNTAX_FALLBACKS,
  SYNTAX_FONT_FACE,
  SYNTAX_FONT_FACE_MIXED,
  SYNTAX_FONT_FACE_MULTIPLE,
  SYNTAX_GLOBAL,
  SYNTAX_KEYFRAMES,
  SYNTAX_MEDIA_QUERY,
  SYNTAX_MULTI_SELECTOR,
  SYNTAX_PROPERTIES,
  SYNTAX_PSEUDO,
  SYNTAX_SUPPORTS,
  SYNTAX_VIEWPORT,
  SYNTAX_CHARSET,
  SYNTAX_IMPORT,
  SYNTAX_IMPORT_MULTIPLE,
} from '../../../tests/mocks';
import { cleanStyles } from '../../../tests/helpers';

describe('JSSAesthetic', () => {
  let instance: JSSAesthetic<any>;

  function renderAndTest(styles: any, expStyles: any, global: boolean = false) {
    const nativeSheet = global
      ? instance.syntax.convertGlobalSheet(styles).toObject()
      : instance.syntax.convertStyleSheet(styles).toObject();
    const styleSheet = instance.processStyleSheet(nativeSheet, 'test');

    expect(nativeSheet).toEqual(expStyles);

    expect(instance.transformStyles(...Object.values(styleSheet))).toMatchSnapshot();

    expect(cleanStyles(instance.sheet.toString())).toMatchSnapshot();
  }

  beforeEach(() => {
    const jss = create();
    jss.setup(preset());

    instance = new JSSAesthetic(jss);
  });

  it('converts and transforms inline styles', () => {
    expect(instance.transformToClassName(['foo', { margin: 0 }, { padding: 2 }])).toMatchSnapshot();
    expect(instance.transformToClassName(['foo', null])).toMatchSnapshot();
  });

  describe('global sheet', () => {
    it('handles globals', () => {
      renderAndTest(
        SYNTAX_GLOBAL,
        {
          '@global': {
            body: { margin: 0 },
            html: { height: '100%' },
            a: {
              color: 'red',
              '&:hover': {
                color: 'darkred',
              },
            },
          },
        },
        true,
      );
    });

    it('handles @font-face', () => {
      renderAndTest(
        SYNTAX_FONT_FACE,
        {
          '@font-face': [FONT_ROBOTO_FLAT_SRC],
        },
        true,
      );
    });

    it('handles mixed @font-face', () => {
      renderAndTest(
        SYNTAX_FONT_FACE_MIXED,
        {
          '@font-face': [FONT_ROBOTO_FLAT_SRC, ...FONT_CIRCULAR_MULTIPLE_FLAT_SRC],
        },
        true,
      );
    });

    it('handles multiple @font-face', () => {
      renderAndTest(
        SYNTAX_FONT_FACE_MULTIPLE,
        {
          '@font-face': FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
        },
        true,
      );
    });

    it('handles @keyframes', () => {
      renderAndTest(
        SYNTAX_KEYFRAMES,
        {
          '@keyframes fade': KEYFRAME_FADE,
        },
        true,
      );
    });

    it('handles @viewport', () => {
      renderAndTest(
        SYNTAX_VIEWPORT,
        {
          '@viewport': {
            width: 'device-width',
            orientation: 'landscape',
          },
        },
        true,
      );
    });

    it('handles @charset', () => {
      renderAndTest(
        SYNTAX_CHARSET,
        {
          '@charset': 'utf8',
        },
        true,
      );
    });

    it('handles single @import', () => {
      renderAndTest(
        SYNTAX_IMPORT,
        {
          '@import': ['./some/path.css'],
        },
        true,
      );
    });

    it('handles multiple @import', () => {
      renderAndTest(
        SYNTAX_IMPORT_MULTIPLE,
        {
          '@import': ['./some/path.css', './another/path.css'],
        },
        true,
      );
    });
  });

  describe('style sheet', () => {
    it('converts unified syntax to native syntax and transforms to a class name', () => {
      renderAndTest(SYNTAX_UNIFIED_LOCAL_FULL, {
        button: {
          margin: 0,
          padding: '6px 12px',
          border: '1px solid #2e6da4',
          borderRadius: 4,
          display: 'inline-block',
          cursor: 'pointer',
          fontFamily: 'Roboto',
          fontWeight: 'normal',
          lineHeight: 'normal',
          whiteSpace: 'nowrap',
          textDecoration: 'none',
          textAlign: 'center',
          backgroundColor: '#337ab7',
          verticalAlign: 'middle',
          color: 'rgba(0, 0, 0, 0)',
          animationName: '$fade',
          animationDuration: '.3s',
          '&:hover': {
            backgroundColor: '#286090',
            borderColor: '#204d74',
          },
          '&::before': {
            content: '"â˜…"',
            display: 'inline-block',
            verticalAlign: 'middle',
            marginRight: 5,
          },
          '@media (max-width: 600px)': {
            padding: '4px 8px',
          },
        },
      });
    });

    it('handles properties', () => {
      renderAndTest(SYNTAX_PROPERTIES, SYNTAX_PROPERTIES);
    });

    it('handles attribute selectors', () => {
      renderAndTest(SYNTAX_ATTRIBUTE, {
        attr: {
          display: 'block',
          '&[disabled]': {
            opacity: 0.5,
          },
        },
      });
    });

    it('handles descendant selectors', () => {
      renderAndTest(SYNTAX_DESCENDANT, {
        list: {
          margin: 0,
          padding: 0,
          '&> li': {
            listStyle: 'bullet',
          },
        },
      });
    });

    it('handles pseudo selectors', () => {
      renderAndTest(SYNTAX_PSEUDO, {
        pseudo: {
          position: 'fixed',
          '&:hover': {
            position: 'static',
          },
          '&::before': {
            position: 'absolute',
          },
        },
      });
    });

    it('handles multiple selectors (comma separated)', () => {
      renderAndTest(SYNTAX_MULTI_SELECTOR, {
        multi: {
          cursor: 'pointer',
          '&:disabled': { cursor: 'default' },
          '&[disabled]': { cursor: 'default' },
          '&> span': { cursor: 'default' },
        },
      });
    });

    it('handles @fallbacks', () => {
      renderAndTest(SYNTAX_FALLBACKS, {
        fallback: {
          background: 'linear-gradient(...)',
          display: 'flex',
          fallbacks: [
            { background: 'red' },
            { display: 'box' },
            { display: 'flex-box' },
            { color: 'blue' },
          ],
        },
      });
    });

    it('handles @media', () => {
      renderAndTest(SYNTAX_MEDIA_QUERY, {
        media: {
          color: 'red',
          '@media (max-width: 1000px)': {
            color: 'green',
          },
          '@media (min-width: 300px)': {
            color: 'blue',
          },
        },
      });
    });

    it('handles @supports', () => {
      renderAndTest(SYNTAX_SUPPORTS, {
        sup: {
          display: 'block',
          '@supports (display: flex)': {
            display: 'flex',
          },
          '@supports not (display: flex)': {
            float: 'left',
          },
        },
      });
    });
  });
});
