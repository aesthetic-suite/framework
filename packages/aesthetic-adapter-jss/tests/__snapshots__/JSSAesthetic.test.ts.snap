// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSSAesthetic converts and transforms inline styles 1`] = `"foo inline-dynamic-inline-0-0-1-1 inline-dynamic-inline-1-0-1-2"`;

exports[`JSSAesthetic converts and transforms inline styles 2`] = `"foo"`;

exports[`JSSAesthetic global sheet handles @charset 1`] = `""`;

exports[`JSSAesthetic global sheet handles @charset 2`] = `"@charset utf8;"`;

exports[`JSSAesthetic global sheet handles @font-face 1`] = `""`;

exports[`JSSAesthetic global sheet handles @font-face 2`] = `"@font-face {font-family: Roboto;font-style: normal;font-weight: normal;src: local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}"`;

exports[`JSSAesthetic global sheet handles @keyframes 1`] = `""`;

exports[`JSSAesthetic global sheet handles @keyframes 2`] = `"@keyframes fade {from {opacity: 0;}to {opacity: 1;}}"`;

exports[`JSSAesthetic global sheet handles @viewport 1`] = `""`;

exports[`JSSAesthetic global sheet handles @viewport 2`] = `"@viewport {width: device-width;orientation: landscape;}"`;

exports[`JSSAesthetic global sheet handles globals 1`] = `""`;

exports[`JSSAesthetic global sheet handles globals 2`] = `"body {margin: 0;}html {height: 100%;}a {color: red;}a:hover {color: darkred;}"`;

exports[`JSSAesthetic global sheet handles mixed @font-face 1`] = `""`;

exports[`JSSAesthetic global sheet handles mixed @font-face 2`] = `"@font-face {font-family: Roboto;font-style: normal;font-weight: normal;src: local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}@font-face {font-family: Circular;font-style: normal;font-weight: normal;src: url('fonts/Circular.woff2') format('woff2');}@font-face {font-family: Circular;font-style: italic;font-weight: normal;src: url('fonts/Circular-Italic.woff2') format('woff2');}@font-face {font-family: Circular;font-style: normal;font-weight: 300;src: url('fonts/Circular-Light.woff2') format('woff2');}@font-face {font-family: Circular;font-style: normal;font-weight: 700;src: url('fonts/Circular-Bold.woff2') format('woff2');}"`;

exports[`JSSAesthetic global sheet handles multiple @font-face 1`] = `""`;

exports[`JSSAesthetic global sheet handles multiple @font-face 2`] = `"@font-face {font-family: Circular;font-style: normal;font-weight: normal;src: url('fonts/Circular.woff2') format('woff2');}@font-face {font-family: Circular;font-style: italic;font-weight: normal;src: url('fonts/Circular-Italic.woff2') format('woff2');}@font-face {font-family: Circular;font-style: normal;font-weight: 300;src: url('fonts/Circular-Light.woff2') format('woff2');}@font-face {font-family: Circular;font-style: normal;font-weight: 700;src: url('fonts/Circular-Bold.woff2') format('woff2');}"`;

exports[`JSSAesthetic global sheet handles multiple @import 1`] = `""`;

exports[`JSSAesthetic global sheet handles multiple @import 2`] = `"@import ./some/path.css;@import ./another/path.css;"`;

exports[`JSSAesthetic global sheet handles single @import 1`] = `""`;

exports[`JSSAesthetic global sheet handles single @import 2`] = `"@import ./some/path.css;"`;

exports[`JSSAesthetic style sheet converts unified syntax to native syntax and transforms to a class name 1`] = `"test-button-0-11-1"`;

exports[`JSSAesthetic style sheet converts unified syntax to native syntax and transforms to a class name 2`] = `".test-button-0-11-1 {color: rgba(0, 0, 0, 0);margin: 0;border: 1px solid #2e6da4;cursor: pointer;display: inline-block;padding: 6px 12px;text-align: center;font-weight: normal;line-height: normal;white-space: nowrap;font-family: Roboto;border-radius: 4px;vertical-align: middle;animation-name: $fade;text-decoration: none;background-color: #337ab7;animation-duration: .3s;}.test-button-0-11-1:hover {border-color: #204d74;background-color: #286090;}.test-button-0-11-1::before {content: \\"â˜…\\";display: inline-block;margin-right: 5px;vertical-align: middle;}@media (max-width: 600px) {.test-button-0-11-1 {padding: 4px 8px;}}"`;

exports[`JSSAesthetic style sheet handles @fallbacks 1`] = `"test-fallback-0-17-1"`;

exports[`JSSAesthetic style sheet handles @fallbacks 2`] = `".test-fallback-0-17-1 {background: red;display: block;display: inline-block;color: blue;display: flex;background: linear-gradient(...);}"`;

exports[`JSSAesthetic style sheet handles @media 1`] = `"test-media-0-18-1"`;

exports[`JSSAesthetic style sheet handles @media 2`] = `".test-media-0-18-1 {color: red;}@media (min-width: 300px) {.test-media-0-18-1 {color: blue;}}@media (max-width: 1000px) {.test-media-0-18-1 {color: green;}}"`;

exports[`JSSAesthetic style sheet handles @supports 1`] = `"test-sup-0-19-1"`;

exports[`JSSAesthetic style sheet handles @supports 2`] = `".test-sup-0-19-1 {display: block;}@supports (display: flex) {.test-sup-0-19-1 {display: flex;}}@supports not (display: flex) {.test-sup-0-19-1 {float: left;}}"`;

exports[`JSSAesthetic style sheet handles attribute selectors 1`] = `"test-attr-0-13-1"`;

exports[`JSSAesthetic style sheet handles attribute selectors 2`] = `".test-attr-0-13-1 {display: block;}.test-attr-0-13-1[disabled] {opacity: 0.5;}"`;

exports[`JSSAesthetic style sheet handles descendant selectors 1`] = `"test-list-0-14-1"`;

exports[`JSSAesthetic style sheet handles descendant selectors 2`] = `".test-list-0-14-1 {margin: 0;padding: 0;}.test-list-0-14-1> li {list-style: bullet;}"`;

exports[`JSSAesthetic style sheet handles multiple selectors (comma separated) 1`] = `"test-multi-0-16-1"`;

exports[`JSSAesthetic style sheet handles multiple selectors (comma separated) 2`] = `".test-multi-0-16-1 {cursor: pointer;}.test-multi-0-16-1:disabled {cursor: default;}.test-multi-0-16-1[disabled] {cursor: default;}.test-multi-0-16-1> span {cursor: default;}"`;

exports[`JSSAesthetic style sheet handles properties 1`] = `"test-props-0-12-1"`;

exports[`JSSAesthetic style sheet handles properties 2`] = `".test-props-0-12-1 {color: black;margin: 10px;display: inline;}"`;

exports[`JSSAesthetic style sheet handles pseudo selectors 1`] = `"test-pseudo-0-15-1"`;

exports[`JSSAesthetic style sheet handles pseudo selectors 2`] = `".test-pseudo-0-15-1 {position: fixed;}.test-pseudo-0-15-1:hover {position: static;}.test-pseudo-0-15-1::before {position: absolute;}"`;
